jina/serve/networking.py

```diff
@@ -670,42 +674,55 @@ async def close(self):
         """
         await self._connections.close()
 
-    def _handle_aiorpcerror(
+    async def _handle_aiorpcerror(
         self,
-        e: AioRpcError,
+        error: AioRpcError,
         retry_i: int = 0,
         request_id: str = '',
-        dest_addr: Set[str] = {''},
-        num_retries: int = 0,
+        tried_addresses: Set[str] = {
+            ''
+        },  # same deployment can have multiple addresses (replicas)
+        total_num_tries: int = 1,  # number of retries + 1
+        current_address: str = '',  # the specific address that was contacted during this attempt
+        connection_list: ReplicaList = None,
     ):
-        # connection failures and cancelled requests should be retried
+        # connection failures, cancelled requests, and timed out requests should be retried
         # all other cases should not be retried and will be raised immediately
         # connection failures have the code grpc.StatusCode.UNAVAILABLE
         # cancelled requests have the code grpc.StatusCode.CANCELLED
+        # timed out requests have the code grpc.StatusCode.DEADLINE_EXCEEDED
         # requests usually gets cancelled when the server shuts down
         # retries for cancelled requests will hit another replica in K8s
         if (
-            e.code() != grpc.StatusCode.UNAVAILABLE
-            and e.code() != grpc.StatusCode.CANCELLED
-            and e.code() != grpc.StatusCode.DEADLINE_EXCEEDED
+            error.code() != grpc.StatusCode.UNAVAILABLE
+            and error.code() != grpc.StatusCode.CANCELLED
+            and error.code() != grpc.StatusCode.DEADLINE_EXCEEDED
         ):
             raise
         elif (
-            e.code() == grpc.StatusCode.UNAVAILABLE
-            or e.code() == grpc.StatusCode.DEADLINE_EXCEEDED
-        ) and retry_i >= num_retries - 1:
+            error.code() == grpc.StatusCode.UNAVAILABLE
+            or error.code() == grpc.StatusCode.DEADLINE_EXCEEDED
+        ) and retry_i >= total_num_tries - 1:  # retries exhausted. if we land here it already failed once, therefore -1
             self._logger.debug(f'GRPC call failed, retries exhausted')
             from jina.excepts import InternalNetworkError
 
+            # after connection failure the gRPC `channel` gets stuck in a failure state for a few seconds
+            # removing and re-adding the connection (stub) is faster & more reliable than just waiting
+            if connection_list:
+                await asyncio.create_task(
```

Comment:
u should be able to simply do `await connection_list.remove_connection(current_address)`